blockchain/chaingen/generator.go

```diff
@@ -1423,6 +1350,239 @@ func (g *Generator) CreateSpendTxForTx(tx *wire.MsgTx, blockHeight, txIndex uint
 	return g.CreateSpendTx(&spend, fee)
 }
 
+// removeTicket removes the passed index from the provide slice of tickets and
+// returns the resulting slice.  This is an in-place modification.
+func removeTicket(tickets []*stakeTicket, index int) []*stakeTicket {
+	copy(tickets[index:], tickets[index+1:])
+	tickets[len(tickets)-1] = nil // Prevent memory leak
+	tickets = tickets[:len(tickets)-1]
+	return tickets
+}
+
+// connectLiveTickets updates the live ticket pool for a new tip block by
+// removing tickets that are now expired from it, removing the passed winners
+// from it, adding any immature tickets which are now mature to it, and
+// resorting it.
+func (g *Generator) connectLiveTickets(blockHash *chainhash.Hash, height uint32, winners, purchases []*stakeTicket) {
+	// Move expired tickets from the live ticket pool to the expired ticket
+	// pool.
+	ticketMaturity := uint32(g.params.TicketMaturity)
+	ticketExpiry := g.params.TicketExpiry
+	for i := 0; i < len(g.liveTickets); i++ {
+		ticket := g.liveTickets[i]
+		liveHeight := ticket.blockHeight + ticketMaturity
+		expireHeight := liveHeight + ticketExpiry
+		if height >= expireHeight {
+			g.liveTickets = removeTicket(g.liveTickets, i)
+			g.expiredTickets = append(g.expiredTickets, ticket)
+			i--
```

Comment:
This idiom always makes me queasy.  I'd prefer to see a comment here describing it.  This is repeated several times throughout the code and I think we want to annotate them all.